File: mf2/java/src/main/java/com/box/l10n/mojito/mf2/Mf2UnlocalizedNumericFunctions.java
package com.box.l10n.mojito.mf2;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.Map;

final class Mf2UnlocalizedNumericFunctions {
    private Mf2UnlocalizedNumericFunctions() {}

    static void registerFormatters(Map<String, Mf2FunctionRegistry.Formatter> formatters) {
        formatters.put("number", Mf2UnlocalizedNumericFunctions::formatNumber);
        formatters.put("percent", Mf2UnlocalizedNumericFunctions::formatPercent);
        formatters.put("integer", Mf2UnlocalizedNumericFunctions::formatInteger);
    }

    private static String formatNumber(Mf2FunctionRegistry.FunctionCall call)
            throws Mf2Exception {
        double value = Mf2FunctionSupport.parseCallDecimal(call, "Number function requires a numeric operand.");
        String formatted = formatDecimalWithMaximumFractionDigits(
                value, maximumFractionDigits(call));
        if (signDisplayAlways(call) && value >= 0.0) {
            formatted = "+" + formatted;
        }
        return appendMinimumFractionDigits(formatted, minimumFractionDigits(call));
    }

    private static String formatPercent(Mf2FunctionRegistry.FunctionCall call)
            throws Mf2Exception {
        double value = Mf2FunctionSupport.parseCallDecimal(call, "Percent function requires a numeric operand.");
        return formatPercentNumber(
                value,
                signDisplayAlways(call),
                minimumFractionDigits(call),
                maximumFractionDigits(call));
    }

    private static String formatInteger(Mf2FunctionRegistry.FunctionCall call)
            throws Mf2Exception {
        double value = Mf2FunctionSupport.parseCallDecimal(call, "Integer function requires a numeric operand.");
        return Mf2PortableFunctions.formatIntegerNumber(
                Mf2FunctionSupport.truncateInteger(value), signDisplayAlways(call));
    }

    static String selectionOperand(
            double value,
            String functionName,
            int minimumFractionDigits,
            Integer maximumFractionDigits) throws Mf2Exception {
        if (functionName.equals("integer")) {
            if (value < -0x1.0p63 || value >= 0x1.0p63) {
                throw Mf2FunctionSupport.badSelector("Integer selector is outside the supported signed-64-bit range.");
            }
            return Long.toString(Mf2FunctionSupport.truncateInteger(value));
        }
        if (functionName.equals("percent")) {
            value = scaledPercent(value);
        }
        if (functionName.equals("number") || functionName.equals("percent")) {
            return appendMinimumFractionDigits(
                    formatDecimalWithMaximumFractionDigits(value, maximumFractionDigits),
                    minimumFractionDigits);
        }
        return Double.toString(value);
    }

    static String formatDecimalNumber(double value, boolean signDisplayAlways, int minimumFractionDigits) {
        String formatted = BigDecimal.valueOf(value).stripTrailingZeros().toPlainString();
        if (signDisplayAlways && value >= 0.0) {
            formatted = "+" + formatted;
        }
        return appendMinimumFractionDigits(formatted, minimumFractionDigits);
    }

    private static double scaledPercent(double value) throws Mf2Exception {
        double scaled = BigDecimal.valueOf(value).movePointRight(2).doubleValue();
        if (!Double.isFinite(scaled)) throw Mf2Exception.badOperand("Scaled percent operand is outside the supported range.");
        return scaled;
    }

    private static String formatPercentNumber(
            double value,
            boolean signDisplayAlways,
            int minimumFractionDigits,
            Integer maximumFractionDigits) throws Mf2Exception {
        String formatted = formatDecimalWithMaximumFractionDigits(scaledPercent(value), maximumFractionDigits);
        if (signDisplayAlways && value >= 0.0) {
            formatted = "+" + formatted;
        }
        return appendMinimumFractionDigits(formatted, minimumFractionDigits) + "%";
    }

    private static String formatDecimalWithMaximumFractionDigits(double value, Integer digits) {
        if (digits == null) {
            return formatDecimalNumber(value, false, 0);
        }
        String formatted = formatFixedFractionDigits(value, digits);
        while (formatted.contains(".") && formatted.endsWith("0")) {
            formatted = formatted.substring(0, formatted.length() - 1);
        }
        if (formatted.endsWith(".")) {
            formatted = formatted.substring(0, formatted.length() - 1);
        }
        return formatted;
    }

    static String formatFixedFractionDigits(double value, int fractionDigits) {
        return BigDecimal.valueOf(value)
                .setScale(fractionDigits, RoundingMode.HALF_UP)
                .toPlainString();
    }

    private static String appendMinimumFractionDigits(String formatted, int minimumFractionDigits) {
        if (minimumFractionDigits == 0) {
            return formatted;
        }
        int dot = formatted.indexOf('.');
        int fractionDigits = dot < 0 ? 0 : formatted.length() - dot - 1;
        StringBuilder output = new StringBuilder(formatted);
        if (fractionDigits == 0) {
            output.append('.');
        }
        for (int index = fractionDigits; index < minimumFractionDigits; index++) {
            output.append('0');
        }
        return output.toString();
    }

    private static int minimumFractionDigits(Mf2FunctionRegistry.FunctionCall call)
            throws Mf2Exception {
        String value = call.optionValue("minimumFractionDigits", null);
        if (value == null) {
            return 0;
        }
        int minimum = Mf2FunctionSupport.parseNonNegativeOption(value, "minimumFractionDigits option must be a non-negative integer.");
        Integer maximum = maximumFractionDigits(call);
        if (maximum != null && minimum > maximum) throw Mf2FunctionSupport.badOption("minimumFractionDigits must not exceed maximumFractionDigits.");
        return minimum;
    }

    private static Integer maximumFractionDigits(Mf2FunctionRegistry.FunctionCall call)
            throws Mf2Exception {
        String value = call.optionValue("maximumFractionDigits", null);
        if (value == null) {
            return null;
        }
        return Mf2FunctionSupport.parseNonNegativeOption(
                value, "maximumFractionDigits option must be a non-negative integer.");
    }

    private static boolean signDisplayAlways(Mf2FunctionRegistry.FunctionCall call)
            throws Mf2Exception {
        return "always".equals(call.optionValue("signDisplay", null));
    }
}

File: mf2/kotlin/src/main/kotlin/com/box/l10n/mojito/mf2/Mf2UnlocalizedNumericFunctions.kt
package com.box.l10n.mojito.mf2

import java.math.BigDecimal
import java.math.RoundingMode
import kotlin.math.truncate

internal object Mf2UnlocalizedNumericFunctions {
    fun registerFormatters(formatters: MutableMap<String, Mf2FunctionFormatter>) {
        formatters["number"] = ::formatNumber
        formatters["percent"] = ::formatPercent
        formatters["integer"] = ::formatInteger
    }

    private fun formatNumber(call: Mf2FunctionCall): String {
        val value = Mf2PortableFunctions.parseCallDecimal(call, "Number function requires a numeric operand.")
        var formatted = formatDecimalWithMaximumFractionDigits(value, maximumFractionDigits(call))
        if (signDisplayAlways(call) && value >= 0.0) formatted = "+$formatted"
        return appendMinimumFractionDigits(formatted, minimumFractionDigits(call))
    }

    private fun formatPercent(call: Mf2FunctionCall): String {
        val value = Mf2PortableFunctions.parseCallDecimal(call, "Percent function requires a numeric operand.")
        val formatted = formatPercentNumber(
            value,
            signDisplayAlways(call),
            minimumFractionDigits(call),
            maximumFractionDigits(call),
        )
        return formatted
    }

    private fun formatInteger(call: Mf2FunctionCall): String {
        val value = Mf2PortableFunctions.parseCallDecimal(call, "Integer function requires a numeric operand.")
        return Mf2PortableFunctions.formatIntegerNumber(
            truncateInteger(value),
            signDisplayAlways(call),
        )
    }

    fun selectionOperand(
        value: Double,
        functionName: String,
        minimumFractionDigits: Int,
        maximumFractionDigits: Int?,
    ): String =
        when (functionName) {
            "integer" -> {
                if (value < -9223372036854775808.0 || value >= 9223372036854775808.0) {
                    throw Mf2Error.badSelector("Integer selector is outside the supported signed-64-bit range.")
                }
                truncateInteger(value).toString()
            }
            "percent" -> appendMinimumFractionDigits(
                formatDecimalWithMaximumFractionDigits(scaledPercent(value), maximumFractionDigits),
                minimumFractionDigits,
            )
            "number" -> appendMinimumFractionDigits(
                formatDecimalWithMaximumFractionDigits(value, maximumFractionDigits),
                minimumFractionDigits,
            )
            else -> value.toString()
        }

    private fun formatDecimalNumber(value: Double, signAlways: Boolean, minimumFractionDigits: Int): String {
        var formatted = BigDecimal.valueOf(value).stripTrailingZeros().toPlainString()
        if (signAlways && value >= 0.0) formatted = "+$formatted"
        return appendMinimumFractionDigits(formatted, minimumFractionDigits)
    }

    private fun scaledPercent(value: Double): Double =
        BigDecimal.valueOf(value).movePointRight(2).toDouble().takeIf { it.isFinite() }
            ?: throw Mf2Error.badOperand("Scaled percent operand is outside the supported range.")

    private fun formatPercentNumber(
        value: Double,
        signAlways: Boolean,
        minimumFractionDigits: Int,
        maximumFractionDigits: Int?,
    ): String {
        var formatted = formatDecimalWithMaximumFractionDigits(scaledPercent(value), maximumFractionDigits)
        if (signAlways && value >= 0.0) formatted = "+$formatted"
        return appendMinimumFractionDigits(formatted, minimumFractionDigits) + "%"
    }

    private fun formatDecimalWithMaximumFractionDigits(value: Double, digits: Int?): String {
        if (digits == null) return formatDecimalNumber(value, false, 0)
        var formatted = formatFixedFractionDigits(value, digits)
        while (formatted.contains(".") && formatted.endsWith("0")) formatted = formatted.dropLast(1)
        if (formatted.endsWith(".")) formatted = formatted.dropLast(1)
        return formatted
    }

    private fun formatFixedFractionDigits(value: Double, fractionDigits: Int): String =
        BigDecimal.valueOf(value).setScale(fractionDigits, RoundingMode.HALF_UP).toPlainString()

    private fun appendMinimumFractionDigits(formatted: String, minimumFractionDigits: Int): String {
        if (minimumFractionDigits == 0) return formatted
        val dot = formatted.indexOf(".")
        val fractionDigits = if (dot < 0) 0 else formatted.length - dot - 1
        val output = StringBuilder(formatted)
        if (fractionDigits == 0) output.append(".")
        for (index in fractionDigits until minimumFractionDigits) output.append("0")
        return output.toString()
    }

    private fun minimumFractionDigits(call: Mf2FunctionCall): Int {
        val minimum = call.optionValue("minimumFractionDigits", null)
            ?.let { Mf2PortableFunctions.parseNonNegativeOption(it, "minimumFractionDigits option must be a non-negative integer.") } ?: 0
        val maximum = maximumFractionDigits(call)
        if (maximum != null && minimum > maximum) throw Mf2Error.badOption("minimumFractionDigits must not exceed maximumFractionDigits.")
        return minimum
    }

    private fun maximumFractionDigits(call: Mf2FunctionCall): Int? =
        call.optionValue("maximumFractionDigits", null)
            ?.let { Mf2PortableFunctions.parseNonNegativeOption(it, "maximumFractionDigits option must be a non-negative integer.") }

    private fun signDisplayAlways(call: Mf2FunctionCall): Boolean =
        call.optionValue("signDisplay", null) == "always"
}
